Dev: CR-1079 Factor out getSurveyUrl

olleharstedt opened the pull request:
See https://limesurvey.atlassian.net/browse/CR-1079

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Survey link generation moved to a new URL service; mailer now emits consistent tokenized survey links and the legacy URL helper is deprecated. Alias handling adjusted to ensure correct language-specific URLs.

* **Documentation**
  * Clarified admin overview, URL/alias selection behavior, and question-cleanup rules; language alias property documented.

* **Tests**
  * Added survey fixture and unit tests validating multi-language URL generation and alias scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Trischi80 committed:
dev: refactored function getSurveyUrl() from model Survey.php to a service class, created a testfile for it

Trischi80 committed:
dev: created testclass for SurveyUrl

Trischi80 committed:
dev: created testclass for SurveyUrl written a comment for it

Trischi80 committed:
dev: finished testclass for SurveyUrl

olleharstedt committed:
PSR-12 fixes with phpcbf

Trischi80 committed:
dev: fixing test for another testsuite on git hub

Trischi80 committed:
Merge remote-tracking branch 'origin/develop' into task/LS-CR-1079_refactore_getSurveyUrl

Trischi80 committed:
dev: code-check failed

olleharstedt commented on tests/unit/models/SurveyUrlServiceTest.php:
```diff
@@ -0,0 +1,92 @@
+use LimeSurvey\Models\Services\SurveyUrl;
+use LSYii_Application;
+
+class SurveyUrlServiceTest extends \ls\tests\TestBaseClass
```
Nicely done :)

olleharstedt commented on application/core/LimeMailer.php:
```diff
@@ -823,9 +823,19 @@ public function getTokenReplacements()
-        $aTokenReplacements["SURVEYURL"] = $survey->getSurveyUrl($language, ["token" => $token]);
+        $surveyUrlCreator = new \LimeSurvey\Models\Services\SurveyUrl($language, ["token" => $token]);
+        $aTokenReplacements["SURVEYURL"] = $surveyUrlCreator->getUrl(
+            $survey->sid,
```
Hm, looking at the usage, would it make more sense to pass the entire survey object? Since it's using multiple parts of it anyway.

Trischi80 replied:
This will take more time. I'm not sure if this is really necessary right now...

olleharstedt replied:
30 min of work or?

Shnoulle replied:
> Hm, looking at the usage, would it make more sense to pass the entire survey object? Since it's using multiple parts of it anyway.

We already have a static var for survey->findByPk

Unsure on the optimisation here.

olleharstedt commented on application/core/LimeMailer.php:
```diff
@@ -823,9 +823,19 @@ public function getTokenReplacements()
             ->createAbsoluteUrl("/optin/participants", array("surveyid" => $this->surveyId, "token" => $token,"langcode" => $language));
         $this->addUrlsPlaceholders("GLOBALOPTINURL");
-        $aTokenReplacements["SURVEYURL"] = $survey->getSurveyUrl($language, ["token" => $token]);
+        $surveyUrlCreator = new \LimeSurvey\Models\Services\SurveyUrl($language, ["token" => $token]);
```
You probably want a `use LimeSurvey\Models\Services\SurveyUrl` statement at top of file to make the code a bit easier on the eye.

Shnoulle commented on application/models/Survey.php:
```diff
@@ -2280,58 +2261,6 @@ function ($s) {
-     * @param bool $preferShortUrl  If true, tries to return the short URL instead of the traditional one.
-     * @return string
-     */
-    public function getSurveyUrl($language = null, $params = [], $preferShortUrl = true)
```
:+1: 

deprecated until 6.X

olleharstedt replied:
No, until 7.x I thought.

olleharstedt replied:
Oh, we mean the same thing. Yes, marked as deprecated during 6.x, then removed at release of 7.0.0

Shnoulle replied:
> Oh, we mean the same thing. Yes, marked as deprecated during 6.x, then removed at release of 7.0.0

No, i made an error … we are already on 6 : then yes : deprecate until 7 (or 8)

Trischi80 committed:
Merge remote-tracking branch 'origin/develop' into task/LS-CR-1079_refactore_getSurveyUrl

Trischi80 committed:
dev: inserted old function and marked it as deprecated

Shnoulle commented:
Question : did [18821: publicurl not used for SURVEYURL](https://bugs.limesurvey.org/view.php?id=18821) issue have a task ?

olleharstedt commented:
> Question : did [18821: publicurl not used for SURVEYURL](https://bugs.limesurvey.org/view.php?id=18821) issue have a task ?

Hm dunno? Is that related to this PR?

Shnoulle commented:
> > > > Question : did [18821: publicurl not used for SURVEYURL](https://bugs.limesurvey.org/view.php?id=18821) issue have a task ?
> > > 
> > > 
> > > Hm dunno? Is that related to this PR?
> > 
> > 
> > getSurveyUrl are related to getSurveyUrl issue : yes.
> 
> OK, what do you mean with task? Our internal Jira system?

I have no idea of you current project : i just see **Task/ls cr 1079** and seems it seems related, i think the fix can be in this task.

Seem it was partially (if there are no alias : i think it fix the issue) 

I just want to avoid such situation : https://github.com/LimeSurvey/LimeSurvey/pull/2851#issuecomment-1578892624 against https://github.com/LimeSurvey/LimeSurvey/commit/b9d27a514b5bd8665603e6522fced67ed54d9c88

I do the fix, but it was fixed by a task after the PR is done …

olleharstedt commented:
This is just a refactoring task, no fix or new feature here.

coderabbitai[bot] committed:
📝 Add docstrings to `task/LS-CR-1079_refactore_getSurveyUrl`

Docstrings generation was requested by @c-schmitz.

The following files were modified:

* `application/controllers/SurveyAdministrationController.php`
* `application/core/LimeMailer.php`
* `application/models/Survey.php`
* `application/models/services/SurveyUrl.php`

These files were ignored:
* `tests/unit/models/SurveyUrlServiceTest.php`

These file types are not supported:
* `tests/data/surveys/limesurvey_survey_931272_SurveyUrl.lss`

c-schmitz committed:
Merge branch 'master' into task/LS-CR-1079_refactore_getSurveyUrl

c-schmitz committed:
Merge branch 'master' into task/LS-CR-1079_refactore_getSurveyUrl

c-schmitz committed:
Fixed issues mentioned by CoderabbitAI

c-schmitz committed:
Fixed phpdoc

c-schmitz committed:
Fixed URL generation

c-schmitz committed:
Fixed URL generation

c-schmitz committed:
Fixed URL generation

c-schmitz committed:
Merge branch 'master' into task/LS-CR-1079_refactore_getSurveyUrl
